cmd/testrunner/sensei.go

```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return nil, err
+	}
+
+	time.Sleep(time.Second)
+
+	mac := createResp.Macaroon
+
+	cred, err := NewMacaroonCredential(mac)
+	if err != nil {
+		return nil, err
+	}
+
+	opts = append(opts,
+		grpc.WithPerRPCCredentials(cred),
+	)
+
+	logger.Infow("Connecting with macaroon")
+
+	var conn *grpc.ClientConn
+	err = tryFunc(
+		func() error {
+			var err error
+			conn, err = grpc.Dial(rpcHost, opts...)
+			if err != nil {
+				return err
+			}
+			return err
+		}, 60)
+	if err != nil {
+		return nil, err
+	}
+
+	return &senseiConnection{
+		alias:  alias,
+		conn:   conn,
+		pubKey: pubkey,
+		client: sensei.NewNodeClient(conn),
+	}, nil
+}
+
+func (l *senseiConnection) PubKey() string {
+	return l.pubKey.String()
+}
+
+func (l *senseiConnection) Close() {
+	l.conn.Close()
+}
+
+func (l *senseiConnection) Connect(key, host string) error {
+	_, err := l.client.ConnectPeer(context.Background(), &sensei.ConnectPeerRequest{
+		NodeConnectionString: fmt.Sprintf("%v@%v", key, host),
+	})
+
+	return err
+}
+
+func (l *senseiConnection) NewAddress() (string, error) {
+	resp, err := l.client.GetUnusedAddress(
+		context.Background(),
+		&sensei.GetUnusedAddressRequest{},
+	)
+	if err != nil {
+		return "", err
+	}
+	return resp.Address, nil
+}
+
+func (l *senseiConnection) OpenChannel(peerKey string, amtSat int64,
+	pushAmtSat int64, private bool) error {
+
+	_, err := l.client.OpenChannels(context.Background(), &sensei.OpenChannelsRequest{
+		Channels: []*sensei.OpenChannelInfo{
+			{
+				AmtSatoshis: uint64(amtSat),
+				Public:      !private,
+
+				// TODO:
+				// NodePubkeyString:   peerKey,
+				// SpendUnconfirmed:   true,
+				// PushSat:            pushAmtSat,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (l *senseiConnection) SetPolicy(chanPoint *lnrpc.ChannelPoint,
+	policy *graphreader.PolicyData) error {
+
+	// Not necessary for sender node.
+	panic("not implemented")
+}
+
+func (l *senseiConnection) ActiveChannels() (int, error) {
+	resp, err := l.client.ListChannels(context.Background(), &sensei.ListChannelsRequest{})
+	if err != nil {
+		return 0, err
+	}
+
+	// TODO: How to get numer of active channels?
+	return len(resp.Channels), nil
+}
+
+func (l *senseiConnection) IsSynced(totalEdges, localChannels int) (bool, error) {
+	// TODO: How to get number of edges received through gossip + local edges?
+
+	panic("not implemented")
+}
+
+func (l *senseiConnection) AddInvoice(amtMsat int64) (string, error) {
+	// Not necessary for sender node.
+	panic("not implemented")
+}
+
+func (l *senseiConnection) SendPayment(invoice string, aliasMap map[string]string) error {
+	panic("not implemented")
+}
+
+func (l *senseiConnection) HasFunds() error {
+	for {
+		resp, err := l.client.GetBalance(context.Background(),
+			&sensei.GetBalanceRequest{})
+		if err != nil {
+			return err
+		}
+
+		// TODO: Is confirmed balance only?
+		if resp.BalanceSatoshis > 0 {
```

Comment:
I updated the GetBalance endpoint to return more information:

uint64 onchain_balance_sats = 1;
uint64 channel_balance_msats = 2;
uint64 channel_outbound_capacity_msats = 3;
uint64 channel_inbound_capacity_msats = 4;
uint64 usable_channel_outbound_capacity_msats = 5;
uint64 usable_channel_inbound_capacity_msats = 6;

Currently sensei has no access to a mempool so it will be only confirmed transactions and transactions that you broadcast.

Reply:
Isn't `channel_balance` the same as `channel_outbound_capacity`?